src/shogun/regression/GLM.cpp

```diff
@@ -0,0 +1,100 @@
+#include <shogun/lib/config.h>
+
+#include <shogun/labels/RegressionLabels.h>
+#include <shogun/lib/SGVector.h>
+#include <shogun/mathematics/Math.h>
+#include <shogun/mathematics/linalg/LinalgNamespace.h>
+#include <shogun/mathematics/linalg/LinalgSpecialPurposes.h>
+
+#include <shogun/regression/GLM.h>
+#include <utility>
+using namespace shogun;
+
+GLM::GLM() : LinearMachine()
+{
+	init();
+}
+float64_t GLM::log_likelihood(
+    const std::shared_ptr<DenseFeatures<float64_t>>& features,
+    const std::shared_ptr<Labels>& label)
+{
+	auto vector_count = features->get_num_vectors();
+	auto feature_count = features->get_num_features();
+	ASSERT(vector_count > 0 && label->get_num_labels() == vector_count)
+
+	SGVector<float64_t> lambda(vector_count);
+	SGVector<float64_t> beta = LinearMachine::get_w();
+	float64_t beta0 = LinearMachine::get_bias();
+	for (auto i = 0; i < vector_count; i++)
+	{
+		SGVector<float64_t> feature_vector = features->get_feature_vector(i);
+		float64_t res = linalg::dot(feature_vector, beta);
+		lambda.set_element(log(1 + std::exp(beta0 + res)), i);
+	}
+	SGVector<float64_t> likelihood(vector_count);
+	SGVector<float64_t> labels = label->get_values();
+	SGVector<float64_t> log_lambda(vector_count);
+
+	for (auto i = 0; i < vector_count; i++)
+		log_lambda.set_element(log(lambda.get_element(i)), i);
+
+	likelihood = linalg::add(
+	    linalg::element_prod(labels, log_lambda), lambda, 1.0, -1.0);
+	return linalg::sum(likelihood);
+}
+
+SGVector<float64_t> GLM::log_likelihood_derivative(
+    const std::shared_ptr<DenseFeatures<float64_t>>& features,
+    const std::shared_ptr<Labels>& label)
+{
+	auto vector_count = features->get_num_vectors();
```

Comment:
it would be good if you could modularise this.
The log likelihood consists of multiple parts, activation, data terms, etc. You could put each of those into a helper method to make it more readable

Reply:
Okay will do that in my next commit thanks:)